src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,278 @@
-            // receiver can assert that both sides are talking about the same solution snapshot and no weird invariant
-            // breaks have occurred.
-            solutionChecksum.WriteTo(writer);
+        // Wait for both the searching and writing tasks to finish.
+        await Task.WhenAll(findAssetsTask, writeAssetsTask).ConfigureAwait(false);
+        cancellationToken.ThrowIfCancellationRequested();
 
-            // special case
-            if (checksums.Length == 0)
-                return;
+        // If we weren't canceled, we better have found and written out all the expected assets.
+        Contract.ThrowIfTrue(foundChecksumCount != checksums.Length);
 
-            Debug.Assert(assetMap != null);
+        return;
 
-            foreach (var (checksum, asset) in assetMap)
+        async Task FindAllAssetsAsync()
+        {
+            await Task.Yield();
+
+            try
+            {
+                await scope.FindAssetsAsync(
+                    assetPath,
+                    checksums,
+                    (checksum, asset) => channel.Writer.TryWrite((checksum, asset)),
+                    cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
             {
-                Contract.ThrowIfNull(asset);
+                // If Something went wrong ensure that we complete the channel so that the writing task will stop.
+                // Also bubble the exception out so that the outer Task.WhenAll will bubble it up.
+                channel.Writer.TryComplete(ex);
+                throw;
+            }
+            finally
+            {
+                // We finished searching for all the checksums, let the writer know.
+                channel.Writer.TryComplete();
+            }
+        }
 
-                var kind = asset.GetWellKnownSynchronizationKind();
-                checksum.WriteTo(writer);
-                writer.WriteInt32((int)kind);
-                serializer.Serialize(asset, writer, context, cancellationToken);
+        async Task WriteAllAssetsToPipeAsync()
+        {
+            await Task.Yield();
 
-                // We flush after each item as that forms a reasonably sized chunk of data to want to then send over the
-                // pipe for the reader on the other side to read.  This allows the item-writing to remain entirely
-                // synchronous without any blocking on async flushing, while also ensuring that we're not buffering the
-                // entire stream of data into the pipe before it gets sent to the other side.
-                await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+            while (await channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                while (channel.Reader.TryRead(out var item))
+                    await WriteSingleAssetToPipeAsync(item.checksum, item.asset, cancellationToken).ConfigureAwait(false);
             }
```

Comment:
this is why channels are nice.  our writing task can wait for items to write in a non-blocking fashion.  then, once there are items it can grab them in a loop, writing them out to the pipe-writer.